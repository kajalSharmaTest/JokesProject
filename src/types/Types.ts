export interface Joke {
  id: string;
  joke: string;
}
export interface IntialState {
  jokes?: Array<Joke>;
  page: number;
  loading: boolean;
  error?: string;
}
export interface Action {
  type: string;
  payload?: Array<Joke>;
}
