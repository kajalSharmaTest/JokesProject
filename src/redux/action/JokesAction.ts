/* Placing all recducer actions at a common place */
export const REQUEST = 'REQUEST';
export const SUCCESS = 'SUCCESS';
export const FAILURE = 'FAILURE';
export const LOAD_MORE = 'LOAD_MORE';
