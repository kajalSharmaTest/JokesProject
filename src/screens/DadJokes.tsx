/* eslint-disable react-native/no-color-literals */
import React, {useEffect, useReducer} from 'react';
import {JokeReducer} from '../redux/reducer/JokeReducer';
import {IntialState, Joke} from '../types/Types';
import {
  REQUEST,
  SUCCESS,
  FAILURE,
  LOAD_MORE,
} from '../redux/action/JokesAction';
import {
  StyleSheet,
  FlatList,
  SafeAreaView,
  View,
  ActivityIndicator,
  Text,
} from 'react-native';
import JokeItem from '../components/JokeItem';
import {fetchJokes} from '../service/JokeService';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center', // vertical allignment
    justifyContent: 'center', // horizontal alignment because flex-direction is row
  },
  loader: {
    ...StyleSheet.absoluteFillObject,
    justifyContent: 'center',
    backgroundColor: 'rgba(0,0,0,0.5)',
  },
  message: {
    padding: 20,
    color: 'black',
    fontWeight: 'bold',
    textAlign: 'center',
  },
});

const initialState: IntialState = {
  jokes: [] as Joke[],
  page: 1,
  loading: false,
  error: undefined,
};

const DadJokes = (): JSX.Element => {
  const [state, dispatch] = useReducer(JokeReducer, initialState);

  const getJokes = async (): Promise<void> => {
    try {
      dispatch({type: REQUEST});
      const response = await fetchJokes(state.page);
      dispatch({type: SUCCESS, payload: response});
    } catch (error) {
      dispatch({type: FAILURE});
    }
  };

  useEffect(() => {
    getJokes();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [state.page]);

  const loadMoreJokes = (): void => {
    if (!state.loading) {
      dispatch({type: LOAD_MORE});
    }
  };

  return (
    <SafeAreaView style={styles.container}>
      {state.error && <Text style={styles.message}>{state.error}</Text>}
      {state.loading && (
        <View style={styles.loader}>
          <ActivityIndicator color="orange" />
        </View>
      )}
      <FlatList<Joke>
        testID="flat-list"
        data={state.jokes}
        keyExtractor={({id}, index) => id + index}
        renderItem={({item}) => <JokeItem joke={item} />}
        onEndReachedThreshold={0}
        onEndReached={() => loadMoreJokes()}
      />
    </SafeAreaView>
  );
};

export default DadJokes;
