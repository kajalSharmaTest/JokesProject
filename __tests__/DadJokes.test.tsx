/* eslint-disable @typescript-eslint/no-explicit-any */
import 'react-native';
import DadJokes from '../src/screens/DadJokes';
import {IntialState, Joke} from '../src/types/Types';
import * as actions from '../src/redux/action/JokesAction';
import renderer from 'react-test-renderer';
import React from 'react';
import {act} from 'react-test-renderer';
import {FlatList} from 'react-native';

const jokeList: Joke[] = [
  {id: '1', joke: 'Test joke 1'},
  {id: '2', joke: 'Test joke 2'},
  {id: '3', joke: 'Test joke 3'},
];

const emptyJokeState: IntialState = {
  jokes: [] as Joke[],
  page: 1,
  loading: false,
  error: undefined,
};

const successWithJokesListState: IntialState = {
  jokes: jokeList,
  page: 1,
  loading: false,
  error: undefined,
};

describe('DadJokes screen test', () => {
  let wrapper: any;
  let testInstance: any;

  // block of code to be executed before each test (render DadJokes with initial empty state, call useEffect to call api)
  beforeEach(() => {
    wrapper = renderer.create(<DadJokes />);
    testInstance = wrapper.root;
    const useReducer = jest.spyOn(React, 'useReducer');
    useReducer.mockImplementationOnce(callback => {
      const success = {
        type: actions.SUCCESS,
        payload: successWithJokesListState.jokes,
      };
      return callback(emptyJokeState, success);
    });
    const useEffect = jest.spyOn(React, 'useEffect');
    useEffect.mockImplementationOnce(f => f());
  });

  it('should render correctly with success response of jokes list', async () => {
    // wait for response and update the component UI
    await act(() => new Promise(setImmediate));
    expect(wrapper).toMatchSnapshot();
  });
  it('should render correctly with with LoadMore', async () => {
    // wait for response and update the component UI
    await act(() => new Promise(setImmediate));
    // verify FlatList is visible in update UI
    expect(testInstance.findByType(FlatList).props.data).toBeTruthy();
    // trigger onEndReached() on flatlist which will call onLoadMore()
    act(() => {
      testInstance.findByType(FlatList).props.onEndReached();
    });
    // wait for response and update the component UI
    await act(() => new Promise(setImmediate));
    // verify update UI
    expect(wrapper).toMatchSnapshot();
  });
});
