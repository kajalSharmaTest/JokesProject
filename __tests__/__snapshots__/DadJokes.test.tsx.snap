// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DadJokes screen test should render correctly with success response of jokes list 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "black",
        "fontWeight": "bold",
        "padding": 20,
        "textAlign": "center",
      }
    }
  >
    Something went wrong. Please try again later.
  </Text>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="flat-list"
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</RCTSafeAreaView>
`;

exports[`DadJokes screen test should render correctly with with LoadMore 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "black",
        "fontWeight": "bold",
        "padding": 20,
        "textAlign": "center",
      }
    }
  >
    Something went wrong. Please try again later.
  </Text>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="flat-list"
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</RCTSafeAreaView>
`;
